api/checks/update.go

```diff
@@ -256,3 +258,77 @@ func getDiffSummary(aeAPI shared.AppEngineAPI, diffAPI shared.DiffAPI, baseRun,
 	}
 	return summary, nil
 }
+
+type pathKeys []string
+
+func (e pathKeys) Len() int      { return len(e) }
+func (e pathKeys) Swap(i, j int) { e[i], e[j] = e[j], e[i] }
+func (e pathKeys) Less(i, j int) bool {
+	return len(strings.Split(e[i], "/")) > len(strings.Split(e[j], "/"))
+}
+
+// collapseDiff collapses a tree of file paths into a smaller tree of folders.
+func collapseDiff(diff shared.ResultsDiff, limit int) shared.ResultsDiff {
+	keys, _ := shared.MapStringKeys(diff)
+	paths := shared.ToStringSlice(collapsePaths(keys, limit))
+	result := make(shared.ResultsDiff)
+	for k, v := range diff {
+		for _, p := range paths {
+			if strings.HasPrefix(k, p) {
+				result.Add(p, v)
+				break
+			}
+		}
+	}
+	return result
+}
+
+// collapseSummary collapses a tree of file paths into a smaller tree of folders.
+func collapseSummary(summary shared.ResultsSummary, limit int) shared.ResultsSummary {
+	keys, _ := shared.MapStringKeys(summary)
+	paths := shared.ToStringSlice(collapsePaths(keys, limit))
+	result := make(shared.ResultsSummary)
+	for k, v := range summary {
+		for _, p := range paths {
+			if strings.HasPrefix(k, p) {
+				result.Add(p, v)
+				break
+			}
+		}
+	}
+	return result
+}
+
+func collapsePaths(keys []string, limit int) mapset.Set {
+	result := shared.NewSetFromStringSlice(keys)
+	// 10 iterations to avoid edge-case infinite looping risk.
+	for i := 0; i < 10 && result.Cardinality() > limit; i++ {
+		sort.Sort(pathKeys(keys))
```

Comment:
Do we really need to sort in each iteration? Assert `issorted` or something instead?

Reply:
We use a set to track what collapsed paths remain after each loop, but we want to iterate those in longest-first order. The alternative here would be to mutate the array as we go, or do a pre-iteration, but I don't think that code complexity would save us much time (we're only talking sorting 30k items up to 10 times).